hi-ml/src/health_ml/run_ml.py

```diff
@@ -339,29 +356,37 @@ def run(self) -> None:
         Driver function to run a ML experiment
         """
         self.setup()
-        self.init_training()
-        if not self.container.run_inference_only:
-            # Backup the environment variables in case we need to run a second training in the unit tests.
-            old_environ = dict(os.environ)
-
-            # do training
-            with logging_section("Model training"):
-                self.run_training()
-
-            # load model checkpoint for custom inference or additional validation step
-            if self.container.has_custom_test_step() or self.container.run_extra_val_epoch:
-                self.load_model_checkpoint()
-
-            # Run extra validation epoch if enabled
-            if self.container.run_extra_val_epoch:
-                with logging_section("Model Validation to save plots on validation set"):
-                    self.run_validation()
-
-            # Kill all processes besides rank 0
-            self.after_ddp_cleanup(old_environ)
-
-        # Run inference on a single device
-        with logging_section("Model inference"):
-            self.run_inference()
-
-        self.run_regression_test()
+        try:
+            self.init_training()
+            if not self.container.run_inference_only:
+                # Backup the environment variables in case we need to run a second training in the unit tests.
+                old_environ = dict(os.environ)
+
+                # do training
+                with logging_section("Model training"):
+                    self.run_training()
+
+                # load model checkpoint for custom inference or additional validation step
+                if self.container.has_custom_test_step() or self.container.run_extra_val_epoch:
+                    self.load_model_checkpoint()
+
+                # Run extra validation epoch if enabled
+                if self.container.run_extra_val_epoch:
+                    with logging_section("Model Validation to save plots on validation set"):
+                        self.run_validation()
+
+                # Kill all processes besides rank 0
+                self.after_ddp_cleanup(old_environ)
+
+            # Run inference on a single device
+            with logging_section("Model inference"):
+                self.run_inference()
+
+            self.run_regression_test()
+
+        finally:
+            if self.azureml_run_for_logging is not None:
+                try:
+                    self.azureml_run_for_logging.complete()
+                except Exception as ex:
+                    logging.error("Failed to complete AzureML run: %s", ex)
```

Comment:
I wonder if we can be more explicit about the exception here - for what reasons would we not be able to mark the run as complete?

Reply:
I don't really know why the run would not complete. But I thought that the run not completing is not a disaster, and I'd rather have the rest of the runner complete despite that failure.